velox/exec/tests/RowContainerTest.cpp

```diff
@@ -363,17 +522,130 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+      std::sort(
+          indexedRows.begin(),
+          indexedRows.end(),
+          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
+            return rowContainer->compareRows(l.second, r.second) < 0;
+          });
+      auto result =
+          createResultVectorForArrayComparison<T>(indexedRows, decoded);
+      assertEqualVectors(expected, result);
+    }
+  }
+
+  template <typename T>
+  VectorPtr createResultVectorForMapComparison(
+      const std::vector<std::pair<vector_size_t, char*>>& indexedRows,
+      const DecodedVector& decoded) {
+    std::vector<
+        std::optional<std::vector<std::pair<T, std::optional<int32_t>>>>>
+        sorted;
+    for (const auto& irow : indexedRows) {
+      if (decoded.isNullAt(irow.first)) {
+        sorted.push_back(std::nullopt);
+      } else {
+        exec::VectorReader<Map<T, int32_t>> reader(&decoded);
+        auto mapView = reader[irow.first];
+        sorted.push_back(
+            {{{mapView.atIndex(0).first, mapView.atIndex(0).second}}});
       }
-      VectorPtr result = vectorMaker.template flatVectorNullable<T>(sorted);
+    }
+    return makeNullableMapVector(sorted);
+  }
+
+  template <typename T>
+  void testCompareMapFloats(TypePtr type, bool ascending, bool nullsFirst) {
+    // NOTE: set 'isJoinBuild' to true to enable nullable sort key in test.
+    auto rowContainer = makeRowContainer({type}, {type}, false);
+    auto lowest = std::numeric_limits<T>::lowest();
+    auto min = std::numeric_limits<T>::min();
+    auto max = std::numeric_limits<T>::max();
+    auto nan = std::numeric_limits<T>::quiet_NaN();
+    auto inf = std::numeric_limits<T>::infinity();
+
+    VectorPtr values = makeNullableMapVector<T, int32_t>(
+        {{{std::nullopt}},
+         {{{max, 1}}},
+         {{{inf, 2}}},
+         {{{0.0, 3}}},
+         {{{nan, 4}}},
+         {{{lowest, 5}}},
+         {{{min, 6}}}});
+    int numRows = values->size();
+
+    SelectivityVector allRows(numRows);
+    DecodedVector decoded(*values, allRows);
+    std::vector<char*> rows(numRows);
+    std::vector<std::pair<vector_size_t, char*>> indexedRows(numRows);
+    for (size_t i = 0; i < numRows; ++i) {
+      rows[i] = rowContainer->newRow();
+      rowContainer->store(decoded, i, rows[i], 0);
+      indexedRows[i] = std::make_pair(i, rows[i]);
+    }
+
+    std::vector<
+        std::optional<std::vector<std::pair<T, std::optional<int32_t>>>>>
+        expectedOrder{
+            {{std::nullopt}},
+            {{{lowest, 5}}},
+            {{{0.0, 3}}},
+            {{{min, 6}}},
+            {{{max, 1}}},
+            {{{inf, 2}}},
+            {{{nan, 4}}}};
+    setupFloatTestExpectedOutput<std::vector<
+        std::optional<std::vector<std::pair<T, std::optional<int32_t>>>>>>(
+        expectedOrder, ascending, nullsFirst);
+    VectorPtr expected = makeNullableMapVector<T, int32_t>(expectedOrder);
+
+    // Verify compare method with two rows as input
```

Comment:
It seems to me that the tests at an API level are similar across Simple, Array, Map and Row. So a common function can be refactored

(My code is at a semantic level and you will have to figure exact syntax)
```
TypePtr t could be REAL(), ARRAY(REAL()), MAP(REAL()), ROW(REAL())
void testRowContainerCompareAPIs(TypePtr t, ....) {
auto createResult(TypePtr t...) -> VectorPtr {
if (t->isPrimitiveType()) {
  return createResultVectorForSimpleComparison<T>(indexedRows, decoded);
} else (t->isArray()) {
  return createResultVectorForArrayComparison<T>(indexedRows, decoded);
} else (t->isMap()) {
  return createResultVectorForMapComparison<T>(...)
}...
};
{
      std::sort(rows.begin(), rows.end(), [&](const char* l, const char* r) {
        return rowContainer->compare(l, r, 0, {nullsFirst, ascending}) < 0;
      });
      VectorPtr result = BaseVector::create(type, numRows, pool_.get());
      rowContainer->extractColumn(rows.data(), numRows, 0, result);
      assertEqualVectors(expected, result);
    }

    // Verify compare method with row and decoded vector as input
    {
      std::sort(
          indexedRows.begin(),
          indexedRows.end(),
          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
            return rowContainer->compare(
                       l.second,
                       rowContainer->columnAt(0),
                       decoded,
                       r.first,
                       {nullsFirst, ascending}) < 0;
          });
      auto result = createResult(type, indexedRows, decoded);
      assertEqualVectors(expected, result);
    }

    // Verify compareRows method with row as input.
    {
      std::sort(
          indexedRows.begin(),
          indexedRows.end(),
          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
            return rowContainer->compareRows(
                       l.second, r.second, {{nullsFirst, ascending}}) < 0;
          });
      auto result = createResult<T>(indexedRows, decoded);
      assertEqualVectors(expected, result);
    }

    // Verify compareRows method with default compare flags.
    if (ascending && nullsFirst) {
      std::sort(
          indexedRows.begin(),
          indexedRows.end(),
          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
            return rowContainer->compareRows(l.second, r.second) < 0;
          });
      auto result =  createResult<T>(indexedRows, decoded);
      assertEqualVectors(expected, result);
    }
```

That would reduce the code quite a bit and make it easier to follow.

Reply:
I've refactored the code more to have a better separation between the etup for each type and the actual tests that are run.